### DISCUSSION TITLE
Separate the domain model and persistence layer

### DISCUSSION THREAD

**Discussion by mohammadhsn** [doctrine/doctrine2#8757] [Support Questions]
Hi,

I have a design-related question, As I know about **DDD** and some sources I read, most people indicate separations between domain and persistence (or infrastructure in general). Most people use Doctrine ORM when they consider the DDD approach in PHP, domain models (or Entities) are a **part of domain**, so they have no idea about persistence.

Let's imagine we have an `Author` domain model that publishes posts, I thought below code:

```PHP
class Post
{
    public function __construct(
        protected Author $author,
        protected PostTitle $title,
        protected PostContent $content,
        protected DateTimeImmutable $publishedAt
    ) {}
}

class Author
{
    public function __construct(
        protected AuthorEmail $email,
        protected array $posts = []
    ) {}

    public function publishNewPost(Post $post)
    {
        $this->posts []= $posts;
    }
}

```
I fact, what I want is I could use the `Post` and `Author` classes as domain models and **also** as doctrine entities.
The problem is doctrine forces me to use `ArrayCollection` instead of `array` so the domain **has to know** about persistence. I'm looking for a design solution that separates all domain model concerns from doctrine concerns.

**Comment by beberlei** [comment]
@mohammadhsn that is not going to be possible 100%, due to other concerns such as lazy loading. but `ArrayCollection` is also its own library so you could consider it independent from Persistence.
